include/rcpputils/filesystem_helper.hpp

```diff
@@ -0,0 +1,195 @@
+//
+// Software License Agreement (BSD License 2.0)
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions
+// are met:
+//
+//  * Redistributions of source code must retain the above copyright
+//    notice, this list of conditions and the following disclaimer.
+//  * Redistributions in binary form must reproduce the above
+//    copyright notice, this list of conditions and the following
+//    disclaimer in the documentation and/or other materials provided
+//    with the distribution.
+//  * Neither the name of the copyright holders nor the names of its
+//    contributors may be used to endorse or promote products derived
+//    from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+// FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+// COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+// INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+// BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+// LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+// ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+//
+// This file is originally from:
+// https://github.com/ros/pluginlib/blob/1a4de29fa55173e9b897ca8ff57ebc88c047e0b3/pluginlib/include/pluginlib/impl/filesystem_helper.hpp
+
+/// Includes std::filesystem and aliases the namespace to `pluginlib::impl::fs`.
+/**
+ * If std::filesystem is not available the necessary functions are emulated.
+ */
+
+#ifndef RCPPUTILS__FILESYSTEM_HELPER_HPP_
+#define RCPPUTILS__FILESYSTEM_HELPER_HPP_
+
+#if defined(_MSC_VER)
+# if _MSC_VER >= 1900
+#  include <experimental/filesystem>
+namespace rcpputils
+{
+namespace fs = std::experimental::filesystem;
+}  // namespace rcpputils
+
+#  define RCPPUTILS__FILESYSYEM_HELPER__HAS_STD_FILESYSTEM
+# endif
+
+#elif defined(__has_include)
+# if __has_include(< filesystem >) && __cplusplus >= 201703L
+#  include <filesystem>
+
+namespace rcpputils
+{
+namespace fs = std::filesystem;
+}  // namespace rcpputils
+
+#  define RCPPUTILS__FILESYSYEM_HELPER__HAS_STD_FILESYSTEM
+// cppcheck-suppress preprocessorErrorDirective
+# elif __has_include(< experimental / filesystem >)
+#  include <experimental/filesystem>
+
+namespace rcpputils
+{
+namespace fs = std::experimental::filesystem;
+}  // namespace rcpputils
+
+#  define RCPPUTILS__FILESYSYEM_HELPER__HAS_STD_FILESYSTEM
+# endif
+#endif
+
+// The standard library does not provide it, so emulate it.
+#ifndef RCPPUTILS__FILESYSYEM_HELPER__HAS_STD_FILESYSTEM
```

Comment:
I understand the desire to use the standard library as much as we can. However, this is expanding the test matrix quite a bit. We have now multiple competing implementation of the same class and, in the end, we can't use more than what provide the emulation layer anyway.

Naive question: wouldn't it be better to just remove lines 44 to 80, use the "emulation layer" everywhere and directly switch to use the real thing once it's actually non-experimental and included in a C++ standard we target?

Reply:
Yes, ultimately the best case scenario would be to use `std::filesystem` everywhere, but macOS is still lagging in it's support of that.  I believe that even with C++17, `std::filesystem` isn't present.

I found that this header (or functionality like it) is used in a few places so that people could essentially "polyfill" filesystem functionality in their code.

Generally, I was hoping to avoid having to maintain our own variant of `std::filesystem`, but it appears that it won't be an option until macOS catches up.

Reply:
One option would be to directly wrap what is available in `rcutils` with some more convenient C++ functionality for the time being, so that we don't have to maintain two complete implementations.

Reply:
Thanks for the explanation - let's keep this aligned with the STL and hopefully we'll be able to remove this at some point.